processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java

```diff
@@ -162,10 +162,11 @@ public GroupByQueryQueryToolChest(
       Map<String, Object> context
   )
   {
+    if (isNestedQueryPushDown(query, groupByStrategy)) {
```

Comment:
I believe that once `isNestedQueryPushDown(query, groupByStrategy)` returns false, it never changes no matter how many times we call it in this method. Since `mergeGroupByResults()` can call internally itself again if `dataSource` is a `queryDataSource` (https://github.com/apache/incubator-druid/pull/5471/files#diff-631de2c7bba9484c3ffd4a58b4cf691eR193), this might be checked again and again redundantly. I would suggest to add new methods which is called in `mergeGroupByResults()` like below:

```
mergeGroupByResults() -> mergeGroupByResultsWithoutPushDown() -> mergeGroupByResultsWithoutPushDown() if needed ...
                     |-> mergeGroupByResultsWithPushDown()
```